# Ahmad Masood
# 20I-1754
# AI-K
# Lab 01

In [63]:
import pandas as pd
import numpy as np

## Reading Files

In [64]:
df= pd.read_csv("Medicine-Updated.csv")
df

Intensity                                          Diagnosis       Type  \
0    400mg/60mg              Pain, menstrual pain and inflammation     Tablet   
1    400mg/60mg                                          Hepatitis     Tablet   
2         150mg                       Pain, Fever and Inflammation     Tablet   
3         500mg                                              Sugar     Tablet   
4         400mg                     bronchitis,bacterial sinusitis     Tablet   
..          ...                                                ...        ...   
100     200/500  Bacterial Infections and Hospital Acquired Pne...  Injection   
101     200/500  Bacterial Infections and Hospital Acquired Pne...  Injection   
102      50/100  Bacterial Infections and Hospital Acquired Pne...  Injection   
103         100  Bacterial Infections and Hospital Acquired Pne...  Injection   
104  50/100/250  Bacterial Infections and Hospital Acquired Pne...  Injection   

            History  Temp      BP  Age Name of Medicine  
0        Depression   101  120/80   16     Arinac Forte  
1    Polluted water    98  120/80   17           Arinax  
2        Depression   101  120/80   18           Ascard  
3          Diabetes    98  140/90   19           Ascare  
4           Allergy    98  120/80   20           Avelox  
..              ...   ...     ...  ...              ...  
100         Allergy    98  120/80   96          Bakacin  
101         Genetic    98  120/80   97          Mikacin  
102         Allergy    98  120/80   98             Mink  
103    Heart attack    98  120/80   99           Sefkin  
104      Depression    98  120/80  100          Zomacin  

[105 rows x 8 columns]

In [102]:
ECG_df=pd.read_csv("ECG200_TRAIN.csv",delimiter="  ")
ECG_df

-1.0000000e+00   5.0205548e-01   5.4216265e-01   7.2238348e-01  \
0              1.0        0.147647        0.804668        0.367771   
1             -1.0        0.316646        0.243199        0.370471   
2             -1.0        1.168874        2.075901        1.760141   
3              1.0        0.648658        0.752026        2.636231   
4              1.0        0.404733        1.280859        2.515148   
..             ...             ...             ...             ...   
94             1.0        0.581277        0.876188        1.042767   
95            -1.0        2.689017        2.708703        2.008381   
96            -1.0        0.197677        0.455417        0.973110   
97             1.0        0.179500        1.038409        1.946421   
98             1.0        0.073124        0.776054        2.181336   

     1.4288852e+00   2.1365158e+00   2.2811490e+00   1.9362737e+00  \
0         0.243894        0.026614       -0.274402        0.096731   
1         1.063738        1.678187        1.759558        1.697717   
2         1.606446        1.949046        1.302842        0.459332   
3         3.455716        2.118157        0.520620       -0.188627   
4         1.299519        1.453432        0.474275       -1.396562   
..             ...             ...             ...             ...   
94        1.796120        2.541399        2.246653        1.500387   
95        2.235800        1.516982        0.029916       -0.561346   
96        1.935956        2.259463        1.741341        1.158296   
97        2.705141        1.670706       -0.101167       -1.578876   
98        3.440325        2.168475        0.497315       -0.924284   

     1.4688900e+00   1.0088451e+00  ...   9.3104294e-01   6.1029836e-01  \
0        -0.747731       -1.609777  ...       -0.533503       -0.400228   
1         1.612159        1.168188  ...        0.764229        0.610621   
2         0.516412        0.852180  ...        0.419006        0.723888   
3         0.780818        0.933775  ...       -0.097869       -0.136787   
4        -0.647081        0.431945  ...        0.376469        0.277811   
..             ...             ...  ...             ...             ...   
94        1.031521        0.382672  ...        1.002770        0.907869   
95       -0.793702       -0.979371  ...       -0.136610       -0.072176   
96        0.418241       -0.071605  ...        0.482452        0.325569   
97       -0.750906        0.175310  ...        0.324323        0.330489   
98       -1.499227       -0.679328  ...       -0.058935       -0.130638   

     6.3889427e-01   6.8467857e-01   5.8323764e-01   6.4052167e-01  \
0         0.176084        1.111768        2.438428        2.734889   
1         0.552900        0.566786        0.604002        0.777068   
2         1.323947        2.136488        1.746597        1.470220   
3        -0.340237       -0.089441       -0.080297       -0.192584   
4         0.225676        0.159091        0.408354        0.540015   
..             ...             ...             ...             ...   
94        0.916457        0.923975        0.767357        0.656223   
95       -0.082738       -0.138468       -0.120396       -0.089411   
96        0.247991        0.184127        0.050358        0.241988   
97        0.111953        0.448948        0.567132        0.136757   
98       -0.347235       -0.177933       -0.060332       -0.347634   

     7.0858515e-01   7.0501088e-01   7.1381545e-01   4.3376464e-01  
0         1.736054        0.036857       -1.265074       -0.208024  
1         0.812345        0.748848        0.818042        0.539347  
2         1.893512        1.256949        0.800407        0.731540  
3        -0.304704       -0.454556        0.314590        0.582190  
4        -0.027791        0.203476        0.346964        0.339185  
..             ...             ...             ...             ...  
94        0.762357        0.501373       -0.333336       -0.524546  
95       -0.243141       -0.

## Giving Columns Names

In [104]:
colName=list(range(1, 98))
colName=[str(i) for i in colName]
ECG_df.set_axis(colName, axis='columns', inplace=True)
ECG_df

1         2         3         4         5         6         7         8  \
0   1.0  0.147647  0.804668  0.367771  0.243894  0.026614 -0.274402  0.096731   
1  -1.0  0.316646  0.243199  0.370471  1.063738  1.678187  1.759558  1.697717   
2  -1.0  1.168874  2.075901  1.760141  1.606446  1.949046  1.302842  0.459332   
3   1.0  0.648658  0.752026  2.636231  3.455716  2.118157  0.520620 -0.188627   
4   1.0  0.404733  1.280859  2.515148  1.299519  1.453432  0.474275 -1.396562   
..  ...       ...       ...       ...       ...       ...       ...       ...   
94  1.0  0.581277  0.876188  1.042767  1.796120  2.541399  2.246653  1.500387   
95 -1.0  2.689017  2.708703  2.008381  2.235800  1.516982  0.029916 -0.561346   
96 -1.0  0.197677  0.455417  0.973110  1.935956  2.259463  1.741341  1.158296   
97  1.0  0.179500  1.038409  1.946421  2.705141  1.670706 -0.101167 -1.578876   
98  1.0  0.073124  0.776054  2.181336  3.440325  2.168475  0.497315 -0.924284   

           9        10  ...        88        89        90        91        92  \
0  -0.747731 -1.609777  ... -0.533503 -0.400228  0.176084  1.111768  2.438428   
1   1.612159  1.168188  ...  0.764229  0.610621  0.552900  0.566786  0.604002   
2   0.516412  0.852180  ...  0.419006  0.723888  1.323947  2.136488  1.746597   
3   0.780818  0.933775  ... -0.097869 -0.136787 -0.340237 -0.089441 -0.080297   
4  -0.647081  0.431945  ...  0.376469  0.277811  0.225676  0.159091  0.408354   
..       ...       ...  ...       ...       ...       ...       ...       ...   
94  1.031521  0.382672  ...  1.002770  0.907869  0.916457  0.923975  0.767357   
95 -0.793702 -0.979371  ... -0.136610 -0.072176 -0.082738 -0.138468 -0.120396   
96  0.418241 -0.071605  ...  0.482452  0.325569  0.247991  0.184127  0.050358   
97 -0.750906  0.175310  ...  0.324323  0.330489  0.111953  0.448948  0.567132   
98 -1.499227 -0.679328  ... -0.058935 -0.130638 -0.347235 -0.177933 -0.060332   

          93        94        95        96        97  
0   2.734889  1.736054  0.036857 -1.265074 -0.208024  
1   0.777068  0.812345  0.748848  0.818042  0.539347  
2   1.470220  1.893512  1.256949  0.800407  0.731540  
3  -0.192584 -0.304704 -0.454556  0.314590  0.582190  
4   0.540015 -0.027791  0.203476  0.346964  0.339185  
..       ...       ...       ...       ...       ...  
94  0.656223  0.762357  0.501373 -0.333336 -0.524546  
95 -0.089411 -0.243141 -0.119710  0.124042  0.273463  
96  0.241988  0.331451 -0.120006  0.042423  0.343293  
97  0.136757  0.444768  0.151050  0.193378  0.451709  
98 -0.347634 -0.447443 -0.066689 -0.178448 -0.256052  

[99 rows x 97 columns]

## Getting Unique Values and their Indexes

In [67]:
uninque_values=ECG_df["1"].unique()

In [68]:
ECG_df["1"].value_counts()

 1.0    69
-1.0    30
Name: 1, dtype: int64

In [69]:
index_1=[]
index_0=[]
for i in range(len(ECG_df["1"])):
    if ECG_df["1"][i]==1.0:
        index_1.append(i)
    else:
        index_0.append(i)

In [70]:
count_unique=ECG_df["1"].value_counts()

In [71]:
len(count_unique)

2

In [72]:
l1=[]
l2=[]
l3=[]
l4=[]

for i in range(len(index_1)):
    if i<17:
        l1.append(index_1[i])
    elif i>16 and i<34:
        l2.append(index_1[i])
    elif i>33 and i<51:
        l3.append(index_1[i])
    elif i>50 and i<69:
        l4.append(index_1[i])

In [73]:
for i in range(len(index_0)):
    if i<7:
        l1.append(index_0[i])
    elif i>6 and i<14:
        l2.append(index_0[i])
    elif i>13 and i<22:
        l3.append(index_0[i])
    elif i>21 and i<30:
        l4.append(index_0[i])

In [74]:
len(l4)

26

## Shuffling

In [75]:
train_index_1= l1+l2+l3
test_index_1=l4

train_index_2= l1+l2+l4
test_index_2=l3

train_index_3= l1+l4+l3
test_index_3=l2

train_index_4= l4+l2+l3
test_index_4=l1

## Getting Values for training and testing

In [76]:
train_data_1=[]
test_data_1=[]

train_data_2=[]
test_data_2=[]

train_data_3=[]
test_data_3=[]

train_data_4=[]
test_data_4=[]

In [77]:
for i in range(len(train_index)):
    train_data_1.append(ECG_df.iloc[train_index[i]].to_list())

for i in range(len(train_index)):
    train_data_2.append(ECG_df.iloc[train_index[i]].to_list())

for i in range(len(train_index)):
    train_data_3.append(ECG_df.iloc[train_index[i]].to_list())
    
for i in range(len(train_index)):
    train_data_4.append(ECG_df.iloc[train_index[i]].to_list())

In [78]:
for i in range(len(test_index)):
    test_data_1.append(ECG_df.iloc[test_index[i]].to_list())
    
for i in range(len(test_index)):
    test_data_2.append(ECG_df.iloc[test_index[i]].to_list())
    
for i in range(len(test_index)):
    test_data_3.append(ECG_df.iloc[test_index[i]].to_list())
    
for i in range(len(test_index)):
    test_data_4.append(ECG_df.iloc[test_index[i]].to_list())

In [79]:
len(test_data_4)

25

## Making Data Frames and Splitting

In [80]:
train_df_1=pd.DataFrame(train_data_1)
train_df_2=pd.DataFrame(train_data_2)
train_df_3=pd.DataFrame(train_data_3)
train_df_4=pd.DataFrame(train_data_4)

In [81]:
x_train_1=train_df_1.iloc[: , 1:]
x_train_2=train_df_2.iloc[: , 1:]
x_train_3=train_df_3.iloc[: , 1:]
x_train_4=train_df_4.iloc[: , 1:]

y_train_1=train_df_1.iloc[: , :1]
y_train_2=train_df_2.iloc[: , :1]
y_train_3=train_df_3.iloc[: , :1]
y_train_4=train_df_4.iloc[: , :1]

In [82]:
test_df_1=pd.DataFrame(test_data_1)
test_df_2=pd.DataFrame(test_data_2)
test_df_3=pd.DataFrame(test_data_3)
test_df_4=pd.DataFrame(test_data_4)

In [83]:
x_test_1=test_df_1.iloc[: , 1:]
x_test_2=test_df_2.iloc[: , 1:]
x_test_3=test_df_3.iloc[: , 1:]
x_test_4=test_df_4.iloc[: , 1:]


y_test_1=test_df_1.iloc[: , :1]
y_test_2=test_df_2.iloc[: , :1]
y_test_3=test_df_3.iloc[: , :1]
y_test_4=test_df_4.iloc[: , :1]

## Importing Libraries

In [112]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split
from sklearn.model_selection import cross_val_score
import warnings
warnings.filterwarnings('ignore')

## Applying Models

In [98]:
knn = KNeighborsClassifier()
knn.fit(x_train_1, y_train_1)
test_data_predicted_1= knn.predict(x_test_1)
print(accuracy_score(test_data_predicted_1, y_test_1))

0.88


In [99]:
knn = KNeighborsClassifier()
knn.fit(x_train_2, y_train_2)
test_data_predicted_2= knn.predict(x_test_2)
print(accuracy_score(test_data_predicted_2, y_test_2))

0.88


In [100]:
knn = KNeighborsClassifier()
knn.fit(x_train_3, y_train_3)
test_data_predicted_3= knn.predict(x_test_3)
print(accuracy_score(test_data_predicted_3, y_test_3))

0.88


In [101]:
knn = KNeighborsClassifier()
knn.fit(x_train_4, y_train_4)
test_data_predicted_4= knn.predict(x_test_4)
print(accuracy_score(test_data_predicted_4, y_test_4))

0.88


## K-Fold 

In [108]:
cross_val_feat=ECG_df.iloc[: , 1:]
cross_val_labels=ECG_df.iloc[: , :1]

In [110]:
model=KNeighborsClassifier()
scores = cross_val_score(model, cross_val_feat, cross_val_labels, cv=4)
scores

array([0.76      , 0.88      , 0.88      , 0.83333333])

In [115]:
np.mean(scores)

0.8383333333333334

## Train Test Split

In [123]:
X=cross_val_feat
y=cross_val_labels

In [124]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.33, random_state=42)

In [125]:
knn = KNeighborsClassifier()
knn.fit(X_train, y_train)
test_data_predicted= knn.predict(X_test)
print(accuracy_score(test_data_predicted, y_test))

0.8181818181818182


## Manual Splitting

In [126]:
ECG_df

1         2         3         4         5         6         7         8  \
0   1.0  0.147647  0.804668  0.367771  0.243894  0.026614 -0.274402  0.096731   
1  -1.0  0.316646  0.243199  0.370471  1.063738  1.678187  1.759558  1.697717   
2  -1.0  1.168874  2.075901  1.760141  1.606446  1.949046  1.302842  0.459332   
3   1.0  0.648658  0.752026  2.636231  3.455716  2.118157  0.520620 -0.188627   
4   1.0  0.404733  1.280859  2.515148  1.299519  1.453432  0.474275 -1.396562   
..  ...       ...       ...       ...       ...       ...       ...       ...   
94  1.0  0.581277  0.876188  1.042767  1.796120  2.541399  2.246653  1.500387   
95 -1.0  2.689017  2.708703  2.008381  2.235800  1.516982  0.029916 -0.561346   
96 -1.0  0.197677  0.455417  0.973110  1.935956  2.259463  1.741341  1.158296   
97  1.0  0.179500  1.038409  1.946421  2.705141  1.670706 -0.101167 -1.578876   
98  1.0  0.073124  0.776054  2.181336  3.440325  2.168475  0.497315 -0.924284   

           9        10  ...        88        89        90        91        92  \
0  -0.747731 -1.609777  ... -0.533503 -0.400228  0.176084  1.111768  2.438428   
1   1.612159  1.168188  ...  0.764229  0.610621  0.552900  0.566786  0.604002   
2   0.516412  0.852180  ...  0.419006  0.723888  1.323947  2.136488  1.746597   
3   0.780818  0.933775  ... -0.097869 -0.136787 -0.340237 -0.089441 -0.080297   
4  -0.647081  0.431945  ...  0.376469  0.277811  0.225676  0.159091  0.408354   
..       ...       ...  ...       ...       ...       ...       ...       ...   
94  1.031521  0.382672  ...  1.002770  0.907869  0.916457  0.923975  0.767357   
95 -0.793702 -0.979371  ... -0.136610 -0.072176 -0.082738 -0.138468 -0.120396   
96  0.418241 -0.071605  ...  0.482452  0.325569  0.247991  0.184127  0.050358   
97 -0.750906  0.175310  ...  0.324323  0.330489  0.111953  0.448948  0.567132   
98 -1.499227 -0.679328  ... -0.058935 -0.130638 -0.347235 -0.177933 -0.060332   

          93        94        95        96        97  
0   2.734889  1.736054  0.036857 -1.265074 -0.208024  
1   0.777068  0.812345  0.748848  0.818042  0.539347  
2   1.470220  1.893512  1.256949  0.800407  0.731540  
3  -0.192584 -0.304704 -0.454556  0.314590  0.582190  
4   0.540015 -0.027791  0.203476  0.346964  0.339185  
..       ...       ...       ...       ...       ...  
94  0.656223  0.762357  0.501373 -0.333336 -0.524546  
95 -0.089411 -0.243141 -0.119710  0.124042  0.273463  
96  0.241988  0.331451 -0.120006  0.042423  0.343293  
97  0.136757  0.444768  0.151050  0.193378  0.451709  
98 -0.347634 -0.447443 -0.066689 -0.178448 -0.256052  

[99 rows x 97 columns]

In [127]:
train_df=ECG_df.iloc[:69]
test_df=ECG_df.iloc[69:]

In [128]:
x_train_m=train_df.iloc[:,1:]
y_train_m=train_df.iloc[:,:1]

x_test_m=test_df.iloc[:,1:]
y_test_m=test_df.iloc[:,:1]

In [129]:
knn.fit(x_train_m, y_train_m)
test_data_predicted_m= knn.predict(x_test_m)
print(accuracy_score(test_data_predicted_m, y_test_m))

0.8333333333333334


# Report

### Manual K fold gives 0.88 Accuracy while built in gives 0.838
### Manual Splitting gives 0.83 Accuracy while built in gives 0.818